### PR TITLE
Add TestPut for performing non-transactional put load on a cluster.

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

**Comment by petermattis** [comment]
@tschottdorf, @bdarnell Trying to run `TestPut` for longer gets an error on one of the puts:

```
TESTS=TestPut TESTTIMEOUT=24h TESTFLAGS="-d 600s" acceptance/run.sh
...
--- FAIL: TestPut (71.04s)
	put_test.go:78: cannot replace lease 1432687383.430399656,0-1432687385.131966829,0 @1:1 with 1432687384.215077639,0-1432687385.215077639,0 @2:2
```

The failure happens after ~60s. Seems fairly repeatable, though I did have it *not* happen on at least one run.

File: acceptance/single_key_test.go [hunk 1]
```diff
@@ -124,12 +124,22 @@ func TestSingleKey(t *testing.T) {
 
 	// Verify that none of the workers encountered an error.
 	var results []result
-	for i := 0; i < *numNodes; i++ {
-		r := <-resultCh
-		if r.err != nil {
-			t.Fatalf("%d: %v", i, r.err)
+	for len(results) != *numNodes {
```
Comment (tamird) [inline acceptance/single_key_test.go]: nit: `len(results) < *numNodes`?

File: acceptance/single_key_test.go [hunk 2]
```diff
@@ -124,12 +124,22 @@ func TestSingleKey(t *testing.T) {
...
+		select {
+		case <-stopper:
+			t.Fatalf("interrupted")
+		case r := <-resultCh:
+			if r.err != nil {
+				t.Fatal(r.err)
+			}
+			results = append(results, r)
+		case <-time.After(delay):
+			log.Infof("%d", atomic.LoadInt64(&expected))
```
Comment (tamird) [inline acceptance/single_key_test.go]: do you need to break out of the loop here?

File: acceptance/put_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		select {
+		case <-stopper:
+			t.Fatalf("interrupted")
+		case err := <-errs:
+			if err != nil {
+				t.Fatal(err)
+			}
+			i++
+		case <-time.After(delay):
+			log.Infof("%d", atomic.LoadInt64(&count))
```
Comment (tamird) [inline acceptance/put_test.go]: ditto re breaking out of the loop

File: acceptance/single_key_test.go (hunk 1 above)
Comment (petermattis) [inline acceptance/single_key_test.go]: Sure. Done.

File: acceptance/single_key_test.go (hunk 2 above)
Comment (petermattis) [inline acceptance/single_key_test.go]: No, this is just printing out stats every second so that you know the test is proceeding. I'll add a comment to make that clearer.

File: acceptance/put_test.go (hunk 3 above)
Comment (petermattis) [inline acceptance/put_test.go]: See other response.

File: acceptance/put_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	r, _ := util.NewPseudoRand()
+	value := util.RandBytes(r, 8192)
+
+	errs := make(chan error, *numNodes)
+	start := time.Now()
+	deadline := start.Add(*duration)
+	var count int64
+	for i := 0; i < *numNodes; i++ {
+		go func(i int) {
```
Comment (tamird) [inline acceptance/put_test.go]: can this just close over `i`? no need for the argument afaict

File: acceptance/put_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+				}
+			}
+			errs <- nil
+		}(i)
+	}
+
+	for i := 0; i < *numNodes; {
+		delay := 1 * time.Second
+		if i > 1 {
+			delay = 1 * time.Hour
```
Comment (tamird) [inline acceptance/put_test.go]: am i reading this right that this test runs for (numNodes - 2) hours?

Comment (tamird) [inline acceptance/put_test.go]: (in the passing case)

File: acceptance/put_test.go (hunk 4 above)
Comment (petermattis) [inline acceptance/put_test.go]: Closures capture variables by reference. If I didn't pass `i` as a parameter then each of the goroutines would be using the same `i`.

File: acceptance/put_test.go (hunk 5 above)
Comment (petermattis) [inline acceptance/put_test.go]: You're reading it wrong. :)

The `delay` is how long to wait between logging test progress. We log progress every second until the first worker exits at which point we set the delay really large in order to not log progress anymore. The code is wrong. It should really read `i > 0`. And in practice it doesn't seem to matter as the workers all exit nearly simultaneously. I think I'll just get rid of that.

File: acceptance/put_test.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+				if _, err := c.Put(fmt.Sprintf("%08d", k), v); err != nil {
+					errs <- err
+					return
+				}
+			}
+			errs <- nil
+		}(i)
+	}
+
+	for i := 0; i < *numNodes; {
```
Comment (tamird) [inline acceptance/put_test.go]: what does this have to do with the number of nodes?

File: acceptance/put_test.go (hunk 4 above)
Comment (tamird) [inline acceptance/put_test.go]: i is unused :(

File: acceptance/single_key_test.go [hunk 7]
```diff
@@ -124,12 +124,24 @@ func TestSingleKey(t *testing.T) {
...
 	// Verify that none of the workers encountered an error.
 	var results []result
-	for i := 0; i < *numNodes; i++ {
-		r := <-resultCh
-		if r.err != nil {
-			t.Fatalf("%d: %v", i, r.err)
+	for len(results) < *numNodes {
+		delay := 1 * time.Second
+		if len(results) > 0 {
+			delay = 1 * time.Hour
```
Comment (tamird) [inline acceptance/single_key_test.go]: is this the same rationale as above? log progress every second until the first worker succeeds and then "stop" logging progress until the test completes?

File: acceptance/put_test.go (hunk 4 above)
Comment (petermattis) [inline acceptance/put_test.go]: Ah, copy-paste from single_node_test.go where `i` is used. Removed. And irritated at github comments where that context wasn't obvious.

File: acceptance/put_test.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		select {
+		case <-stopper:
+			t.Fatalf("interrupted")
+		case err := <-errs:
+			if err != nil {
+				t.Fatal(err)
+			}
+			i++
```
Comment (tamird) [inline acceptance/put_test.go]: isn't this unreachable?

Comment (tamird) [inline acceptance/put_test.go]: wait, nevermind. so this test terminates when `*numNodes` errors are observed. LGTM

Comment (petermattis) [inline acceptance/put_test.go]: Why unreachable? `err` is `nil` if the worker exited successfully.

File: acceptance/single_key_test.go (hunk 7 above)
Comment (petermattis) [inline acceptance/single_key_test.go]: Yes.

File: acceptance/put_test.go (hunk 6 above)
Comment (petermattis) [inline acceptance/put_test.go]: It is waiting for the `*numNodes` workers to complete.

File: storage/range.go [hunk 9]
```diff
@@ -375,9 +375,13 @@ func (r *Range) redirectOnOrAcquireLeaderLease(timestamp proto.Timestamp) error
 		return r.newNotLeaderError()
 	} else if !held || expired {
 		// Otherwise, if not held by this replica or expired, request renewal.
-		if err := r.requestLeaderLease(timestamp); err != nil {
-			return err
+		err := r.requestLeaderLease(timestamp)
+		// Getting a LeaseRejectedError back means someone else got there
+		// first. Since it's committed by now, we can return a NotLeaderError.
+		if _, ok := err.(*proto.LeaseRejectedError); ok {
```
Comment (vivekmenezes) [inline storage/range.go]: err == nil ?

Comment (petermattis) [inline storage/range.go]: This was pulled in from @tschottdorf's change. I've since reverted it.

Comment (tschottdorf) [inline storage/range.go]: if err is nil, then `ok = false`. If an `interface{}(*proto.LeaseRejectedError(nil))` got returned instead (which might be what you mean), we could check that `err.(*proto.LeaseRejectedError) != nil`, but a) that does not happen and b) even if it happened, we'd probably still do the same thing here.

Comment (vivekmenezes) [inline storage/range.go]: if err can never be nil we should just panic on it. As a reader I'd assume r.requestLeaderLease(timestamp) can return nil.

Its really annoying that golang doesn't allow you to throw invariants all around in the code. modeling it as

if invariant {
panic(..)
}

is ugly

Comment (tschottdorf) [inline storage/range.go]: I probably wasn't clear enough above. `err` is almost always nil, and only if it isn't AND it's pointing at a `*proto.LeaseRejectedError` we are entering the `if`.
